pkg/prometheus/statefulset.go

```diff
@@ -1076,3 +676,488 @@ func queryLogFilePath(queryLogFile string) string {
+	rules monitoringv1.Rules,
+	query *monitoringv1.QuerySpec,
+	allowOverlappingBlocks bool) {
+	var (
+		retentionTimeFlagName  = "storage.tsdb.retention.time"
+		retentionTimeFlagValue = string(retention)
+	)
+	if cg.WithMaximumVersion("2.7.0").IsCompatible() {
+		retentionTimeFlagName = "storage.tsdb.retention"
+		if retention == "" {
+			retentionTimeFlagValue = defaultRetention
+		}
+	} else if retention == "" && retentionSize == "" {
+		retentionTimeFlagValue = defaultRetention
+	}
+
+	if retentionTimeFlagValue != "" {
+		*promArgs = append(*promArgs, monitoringv1.Argument{Name: retentionTimeFlagName, Value: retentionTimeFlagValue})
+	}
+	if retentionSize != "" {
+		retentionSizeFlag := monitoringv1.Argument{Name: "storage.tsdb.retention.size", Value: string(retentionSize)}
+		*promArgs = cg.WithMinimumVersion("2.7.0").AppendCommandlineArgument(*promArgs, retentionSizeFlag)
+	}
+
+	*promArgs = append(*promArgs,
+		monitoringv1.Argument{Name: "storage.tsdb.path", Value: storageDir},
+	)
+
+	if rules.Alert.ForOutageTolerance != "" {
+		*promArgs = cg.WithMinimumVersion("2.4.0").AppendCommandlineArgument(*promArgs, monitoringv1.Argument{Name: "rules.alert.for-outage-tolerance", Value: rules.Alert.ForOutageTolerance})
+	}
+	if rules.Alert.ForGracePeriod != "" {
+		*promArgs = cg.WithMinimumVersion("2.4.0").AppendCommandlineArgument(*promArgs, monitoringv1.Argument{Name: "rules.alert.for-grace-period", Value: rules.Alert.ForGracePeriod})
+	}
+	if rules.Alert.ResendDelay != "" {
+		*promArgs = cg.WithMinimumVersion("2.4.0").AppendCommandlineArgument(*promArgs, monitoringv1.Argument{Name: "rules.alert.resend-delay", Value: rules.Alert.ResendDelay})
+	}
+
+	if query != nil {
+		if query.LookbackDelta != nil {
+			*promArgs = append(*promArgs, monitoringv1.Argument{Name: "query.lookback-delta", Value: *query.LookbackDelta})
+		}
+
+		if query.MaxSamples != nil && *query.MaxSamples > 0 {
+			*promArgs = cg.WithMinimumVersion("2.5.0").AppendCommandlineArgument(*promArgs, monitoringv1.Argument{Name: "query.max-samples", Value: fmt.Sprintf("%d", *query.MaxSamples)})
+		}
+
+		if query.MaxConcurrency != nil && *query.MaxConcurrency > 1 {
+			*promArgs = append(*promArgs, monitoringv1.Argument{Name: "query.max-concurrency", Value: fmt.Sprintf("%d", *query.MaxConcurrency)})
+		}
+
+		if query.Timeout != nil {
+			*promArgs = append(*promArgs, monitoringv1.Argument{Name: "query.timeout", Value: string(*query.Timeout)})
+		}
+	}
+
+	if allowOverlappingBlocks {
+		*promArgs = cg.WithMinimumVersion("2.11.0").WithMaximumVersion("2.39.0").AppendCommandlineArgument(*promArgs, monitoringv1.Argument{Name: "storage.tsdb.allow-overlapping-blocks"})
+	}
+}
+
+// AddCommonVolumes adds volumes to the statefulset spec that are common between Prometheus Server and Agent
+func buildCommonVolumes(cpf monitoringv1.CommonPrometheusFields, promName, volName string, tlsAssetSecrets []string) ([]v1.Volume, []v1.VolumeMount, error) {
+	assetsVolume := v1.Volume{
```

Comment:
And I suspect we might also be able to stop passing `volName` 👀 but didn't look deep into it
```suggestion
func buildCommonVolumes(p monitoringv1.PrometheusInterface, volName string, tlsAssetSecrets []string) ([]v1.Volume, []v1.VolumeMount, error) {
        cpf := p.GetCommonPrometheusFields()
        promName := p.GetObjectMeta().GetName()
	assetsVolume := v1.Volume{
```

Reply:
We could get rid of volName, but it's needed again when creating the Thanos container.

Reply:
Looking at the createThanosContainer signature, we could replace 2 arguments with one `p PrometheusInterface`, maybe it's worth reducing signature for better readability 🙂